sqlalchemy_bigquery/merge.py

```diff
@@ -0,0 +1,231 @@
+# Copyright 2021 Google LLC
+#
+# Use of this source code is governed by an MIT-style
+# license that can be found in the LICENSE file or at
+# https://opensource.org/licenses/MIT.
+
+import textwrap
+from copy import copy
+from dataclasses import dataclass
+from typing import Any, ClassVar, Dict, Optional, Tuple, Type, Union
+
+import sqlalchemy as sa
+from sqlalchemy.ext.compiler import compiles
+from sqlalchemy.sql.compiler import SQLCompiler
+from sqlalchemy.sql.expression import ClauseElement, ColumnElement
+from sqlalchemy.sql.selectable import Subquery  # noqa
+from sqlalchemy.sql.selectable import Alias, Select, TableClause
+
+MergeConditionType = Optional[ColumnElement[sa.Boolean]]
+
+
+class Merge(ClauseElement):
+    def __init__(
+        self,
+        target: Union[Alias, TableClause],
+        source: Union[Alias, TableClause, Subquery, Select],
+        *,
+        on: ColumnElement[sa.Boolean],
+    ):
+        super().__init__()
+        if not (
+            isinstance(target, TableClause)
+            or (
+                isinstance(target, Alias)
+                and isinstance(getattr(target, "element"), TableClause)
+            )
+        ):
+            raise Exception(
+                "Parameter `target` must be a table, or an aliased table,"
+                f" instead received:\n{repr(target)}"
+            )
+        if not isinstance(target, (Alias, TableClause, Subquery, Select)):
+            raise Exception(
+                "Parameter `source` must be a table, alias, subquery, or selectable,"
+                f" instead received:\n{repr(target)}"
+            )
+
+        self.when = tuple()
+        self.target = target
+        self.source = source
+        self.on = on
+        self.when: Tuple[WhenBase] = tuple()
+
+    def when_matched(self, condition: MergeConditionType = None):
+        return MergeWhenMatched(self, condition)
+
+    def when_matched_not_matched_by_target(self, condition: MergeConditionType = None):
+        return MergeWhenNotMatchedByTarget(self, condition)
+
+    def when_matched_not_matched_by_source(self, condition: MergeConditionType = None):
+        return MergeWhenNotMatchedBySource(self, condition)
```

Comment:
Shouldn't they be `when_not_matched_by_target` and `when_not_matched_by_source`? 🤔